File: Data Structures & Algorithms/daily-temperatures/submission-2.java
class Solution {
    public record Pair(int tmp, int index) {}

    public int[] dailyTemperatures(int[] temperatures) {
        int[] result = new int[temperatures.length];
        Stack<Pair> stack = new Stack<>();
        for (int i = 0; i < temperatures.length; i++) {
            while (!stack.empty() && stack.peek().tmp() < temperatures[i]) {
                Pair top = stack.peek();
                result[top.index] = i - top.index();
                stack.pop();
            }
            Pair newP = new Pair(temperatures[i], i);
            stack.push(newP);
        }
        return result;
    }
}

File: Data Structures & Algorithms/jump-game/submission-2.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        memo = {}
        n = len(nums) -1
        def dp(index):
            if index >= n:
                return True
            if index in memo:
                return memo[index]
        
            result = False
            if nums[index] == 0:
                memo[index] = result
                return result
            
            for i in range(nums[index], 0, -1):
                if dp(index + i):
                    result = True
                    break

            memo[index] = result
            return result
        return dp(0)

